apprise/plugins/NotifyEmail.py

```diff
@@ -778,58 +786,79 @@ def send(self, body, title='', notify_type=NotifyType.INFO, attach=None,
             if reply_to:
                 base['Reply-To'] = ','.join(reply_to)
 
-            # bind the socket variable to the current namespace
-            socket = None
+            message = EmailMessage(
+                recipient=to_addr,
+                to_addrs=[to_addr] + list(cc) + list(bcc),
+                body=base.as_string())
+            messages.append(message)
 
-            # Always call throttle before any remote server i/o is made
-            self.throttle()
+        return self.submit(messages)
 
-            try:
-                self.logger.debug('Connecting to remote SMTP server...')
-                socket_func = smtplib.SMTP
-                if self.secure and self.secure_mode == SecureMailMode.SSL:
-                    self.logger.debug('Securing connection with SSL...')
-                    socket_func = smtplib.SMTP_SSL
+    def submit(self, messages: t.List[EmailMessage]):
 
-                socket = socket_func(
-                    self.smtp_host,
-                    self.port,
-                    None,
-                    timeout=self.socket_connect_timeout,
-                )
+        # error tracking (used for function return)
+        has_error = False
 
-                if self.secure and self.secure_mode == SecureMailMode.STARTTLS:
-                    # Handle Secure Connections
-                    self.logger.debug('Securing connection with STARTTLS...')
-                    socket.starttls()
+        # bind the socket variable to the current namespace
+        socket = None
 
-                if self.user and self.password:
-                    # Apply Login credetials
-                    self.logger.debug('Applying user credentials...')
-                    socket.login(self.user, self.password)
+        # Always call throttle before any remote server i/o is made
+        self.throttle()
 
-                # Send the email
-                socket.sendmail(
-                    self.from_addr,
-                    [to_addr] + list(cc) + list(bcc),
-                    base.as_string())
+        try:
+            self.logger.debug('Connecting to remote SMTP server...')
+            socket_func = smtplib.SMTP
+            if self.secure and self.secure_mode == SecureMailMode.SSL:
+                self.logger.debug('Securing connection with SSL...')
+                socket_func = smtplib.SMTP_SSL
 
-                self.logger.info(
-                    'Sent Email notification to "{}".'.format(to_addr))
+            socket = socket_func(
+                self.smtp_host,
+                self.port,
+                None,
+                timeout=self.socket_connect_timeout,
+            )
 
-            except (SocketError, smtplib.SMTPException, RuntimeError) as e:
```

Comment:
Does your new implementation no handle these errors (disconnects, timeouts, etc)?

Reply:
Hi Chris,

the new implementation is not so new at all. I effectively only dedented the block where the plugins starts communicating with SMTP.

https://github.com/caronc/apprise/blob/67070e5932a2716b282f297290e095732bedba16/apprise/plugins/NotifyEmail.py#L787-L827

But you are right, of course: I may have missed to reflect error handling appropriately for both connection setup vs. email submission code paths. I will have another look, thank you.

With kind regards,
Andreas.

Reply:
Ah, the error handling of the inner mail submission loop looks like this now:

https://github.com/caronc/apprise/blob/2075df219f0ca7097dedb805e8e57043498a0bbb/apprise/plugins/NotifyEmail.py#L842-L844

Which is all embedded into the error handling already present beforehand:

https://github.com/caronc/apprise/blob/2075df219f0ca7097dedb805e8e57043498a0bbb/apprise/plugins/NotifyEmail.py#L849-L853

I thought it would have been ok like this. Please let me know which improvements you would like to see.

Reply:
I think I'll ask that you to roll this specific area back if you don't mind?

I don't mind the improvement you made to re-use the established connection, but I've always been taught to avoid `Exception` references. I think there is only 1 other notification that violates this (StreamLabs) that needs to be cleaned up. It looks cleaner with just the `Exception` Reference, but i personally like seeing what is being thrown and what except block it's entering into. Makes it easier to separate it later and provide cleaner logs later on (if un-grouped)

I'm also trying to remove any privacy issues from logging (unless  `DEBUG` or `TRACE` set)  There are lots of places that still need some cleanup though; but this was one plugin that wasn't too bad for it.  Also the exception itself definitely has to be hidden in a `self.logger.debug` for sure `warning`).

On a side note... thank you for all the attention you've been giving Apprise these past few days.  I really do appreciate it.

One note with the old way is while it's definitely not making proper uses of the resources, the new way if a disconnect occurs, the exception is thrown all the wy up and it's done.  This may be a good idea or not (i'd be curious on your own views here).  My take is that in the past it would flag that there was an error during the emailing, but would at least try to send out all it could.

Reply:
> I don't mind the improvement you made to re-use the established connection.

Thank you.

> But I've always been taught to avoid `Exception` references.

Oh, I see what you mean. Sure. Fixed.

> [With the] new way, if a disconnect occurs, the exception is thrown all the way up and it's done. This may be a good idea or not - I'd be curious on your own views here.

It is true that the new code will bail out if any exceptions occur during _connection setup_. However, when looping through the email recipients, it will still be graceful and will only issue warnings.

> With the old way, it is definitely not making proper use of the resources.

I think it will definitively lead to being flagged by many SMTP servers, when submitting a reasonable amount of emails to multiple recipients. I know Apprise has a throttling feature, but I think this should be used for orthogonal features only, and not for this particular case.

I think the new code perfectly follows the [principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) with respect to that a user/developer/system operator will never expect that, when submitting email to an SMTP server, addressed to multiple recipients, using an arbitrary program or system, **a new SMTP connection will be made _per recipient/addressee_**.

> My take is that in the past it would flag that there was an error during the emailing, but would at least try to send out all it could.

I see your point. But still, the code doesn't do the opposite: When an error occurs during the submission of a particular email out of multiple ones, for example because of an intermittent network error, it will still continue trying to deliver subsequent emails. Only when an exception happens during _connection setup_, it will bail out. Personally, I think this is perfectly fine, because any kind of "retry functionality" would have to be implemented by upper layers of software, i.e. the one who is _invoking_ the notifier, or another subsystem in between.

> On a side note - thank you for all the attention you've been giving Apprise these past few days. I really do appreciate it.

Thank _you_ for being so responsive, I also appreciate that very much. I hope not to become too much of a pain with any kinds of upcoming issues or patches I might submit.